### PR TITLE
ref/fix: more robust signature verification

### PR DESCRIPTION
**Pull request by mdtro**
Fix for H1 [2547600](https://hackerone.com/reports/2547600).

Increase the robustness of the GitHub Enterprise integration's webhook signature handling.

- Add support for SHA256 signature headers
- Verify the signature value looks as expected before continuing
- Support for legacy GitHub Enterprise installations that omit the header by setting their host value in an option (not the most secure, but we will deprecate use of this legacy method very soon)

### CODE REVIEW

File: src/sentry/integrations/github_enterprise/webhook.py [hunk 1]
```diff
@@ -120,28 +147,35 @@
...
 
-        body = bytes(request.body)
-        if not body:
+        try:
+            body = bytes(request.body)
+        except MissingWebhookPayloadError as e:
             logger.warning("github_enterprise.webhook.missing-body", extra=extra)
-            return HttpResponse(status=400)
+            sentry_sdk.capture_exception(e)
+            return HttpResponse(str(e), status=400)
```
[the remark this answers is not in the document]
Comment (mdtro) [reply src/sentry/integrations/github_enterprise/webhook.py]: Switched this out to use a static string.

File: src/sentry/integrations/github_enterprise/webhook.py [hunk 2]
```diff
@@ -120,28 +147,35 @@
...
+            if not github_event:
+                raise MissingRequiredHeaderError("Missing X-GitHub-Event header")
+
+            handler = self.get_handler(github_event)
+        except MissingRequiredHeaderError as e:
             logger.warning("github_enterprise.webhook.missing-event", extra=extra)
-            logger.exception("Missing Github event in webhook.")
-            return HttpResponse(status=400)
+            sentry_sdk.capture_exception(e)
+            return HttpResponse(str(e), status=400)
```
[the remark this answers is not in the document]
Comment (mdtro) [reply src/sentry/integrations/github_enterprise/webhook.py]: Switched this out to use a static string.

File: src/sentry/integrations/github_enterprise/webhook.py [hunk 3]
```diff
@@ -165,16 +199,76 @@
...
+                if not is_valid:
+                    raise InvalidSignatureError(
+                        "Provided signature does not match the computed body signature"
+                    )
+
+        except InvalidSignatureError as e:
+            logger.warning("github_enterprise.webhook.invalid-signature", extra=extra)
+            sentry_sdk.capture_exception(e)
+
+            return HttpResponse(str(e), status=401)
```
[the remark this answers is not in the document]
Comment (mdtro) [reply src/sentry/integrations/github_enterprise/webhook.py]: Switched this out to use a static string.

File: src/sentry/integrations/github_enterprise/webhook.py [hunk 4]
```diff
@@ -165,16 +199,76 @@
...
+
+            allowed_legacy_hosts = options.get(
+                "github-enterprise-app.allowed-hosts-legacy-webhooks"
+            )
+
+            if host not in allowed_legacy_hosts:
+                # the host is not allowed to skip signature verification by omitting the headers
+                logger.warning("github_enterprise.webhook.missing-signature", extra=extra)
+                sentry_sdk.capture_exception(e)
+                return HttpResponse(str(e), status=400)
```
[the remark this answers is not in the document]
Comment (mdtro) [reply src/sentry/integrations/github_enterprise/webhook.py]: Switched this out to use a static string.

File: src/sentry/integrations/github_enterprise/webhook.py [hunk 5]
```diff
@@ -165,16 +199,76 @@
...
+            else:
+                # the host is allowed to skip signature verification
+                logger.info("github_enterprise.webhook.allowed-missing-signature", extra=extra)
+
+        except (MalformedSignatureError, IndexError) as e:
+            logger.warning("github_enterprise.webhook.malformed-signature", extra=extra)
+            sentry_sdk.capture_exception(e)
+            return HttpResponse(str(e), status=400)
```
[the remark this answers is not in the document]
Comment (mdtro) [reply src/sentry/integrations/github_enterprise/webhook.py]: Switched this out to use a static string.

File: src/sentry/integrations/github_enterprise/webhook.py [hunk 6]
```diff
@@ -165,16 +210,86 @@
...
+                if not is_valid:
+                    raise InvalidSignatureError(
+                        "Provided signature does not match the computed body signature"
+                    )
+
+        except InvalidSignatureError as e:
+            logger.warning("github_enterprise.webhook.invalid-signature", extra=extra)
+            sentry_sdk.capture_exception(e)
+
+            return HttpResponse(str(e), status=401)
```
[the remark this answers is not in the document]
Comment (mdtro) [reply src/sentry/integrations/github_enterprise/webhook.py]: Switched this out to use a static string.

File: src/sentry/integrations/github_enterprise/webhook.py [hunk 7]
```diff
@@ -165,16 +215,78 @@
...
+            return HttpResponse(INVALID_SIGNATURE_ERROR, status=401)
+        except UnsupportedSignatureAlgorithmError as e:
+            # we should never end up here with the regex checks above on the signature format,
+            # but just in case
+            logger.exception(
+                "github-enterprise-app.webhook.unsupported-signature-algorithm",
+                extra=extra,
+            )
+            sentry_sdk.capture_exception(e)
+            return HttpResponse(str(e), 400)
```
[the remark this answers is not in the document]
Comment (mdtro) [reply src/sentry/integrations/github_enterprise/webhook.py]: Switched this out to use a static string.

File: src/sentry/integrations/github_enterprise/webhook.py [hunk 8]
```diff
@@ -120,28 +147,35 @@
...
-        if not host:
+        host = None
+        try:
+            host = get_host(request=request)
+        except MissingRequiredHeaderError as e:
             logger.warning("github_enterprise.webhook.missing-enterprise-host")
-            logger.error("Missing enterprise host.")
-            return HttpResponse(status=400)
+            sentry_sdk.capture_exception(e)
+            return HttpResponse(str(e), status=400)
```
[the remark this answers is not in the document]
Comment (mdtro) [reply src/sentry/integrations/github_enterprise/webhook.py]: Switched this out to use a static string.

**Review by oioki — APPROVED** [review]

**Pull request merged**